src/handlers/rate-limiter.js

```diff
@@ -0,0 +1,151 @@
+      ctx.waitUntil(accounting.record(dataCid, req))
+      return handler(req, env, ctx)
+    }
+  }
+}
+
+/**
+ * @param {Environment} env
+ * @param {IpfsUrlContext} ctx
+ * @returns {RateLimitService}
+ */
+function createRateLimitService (env, ctx) {
+  return {
+    /**
+     * @param {import('multiformats/cid').CID} cid
+     * @param {Request} request
+     * @returns {Promise<RateLimitExceeded>}
+     */
+    check: async (cid, request) => {
+      const authToken = await getAuthorizationTokenFromRequest(request)
+      if (!authToken) {
+        // no token, use normal rate limit
+        return isRateLimited(env.RATE_LIMITER, cid)
+      }
+
+      const tokenMetadata = await getTokenMetadata(env, authToken, ctx)
+      if (!tokenMetadata) {
+        // we didn't get any metadata - for now just use the top level rate limit
+        // this means token based requests will be subject to normal rate limits until the data propagates
+        return isRateLimited(env.RATE_LIMITER, cid)
+      }
+
+      if (tokenMetadata.invalid) {
+        // this means we know about the token and we know it's invalid,
+        // so we should just use the CID rate limit
+        return isRateLimited(env.RATE_LIMITER, cid)
+      }
+
+      // TODO at some point we should enforce user configurable rate limits
+      // and origin matching but for now we just serve all valid token requests
+      return RATE_LIMIT_EXCEEDED.NO
+    }
+  }
+}
+
+/**
+ * @param {Request} request
+ * @returns {Promise<string | null>}
+ */
+async function getAuthorizationTokenFromRequest (request) {
+  const authHeader = request.headers.get('Authorization')
+  if (authHeader && authHeader.startsWith('Bearer ')) {
+    return authHeader.substring(7) // Remove 'Bearer ' prefix
+  }
+  return null
+}
+
+/**
+ * @param {import('@cloudflare/workers-types').RateLimit} rateLimitAPI
+ * @param {import('multiformats/cid').CID} cid
+ * @returns {Promise<import('../constants.js').RATE_LIMIT_EXCEEDED>}
+ */
+async function isRateLimited (rateLimitAPI, cid) {
+  if (!rateLimitAPI) {
+    return RATE_LIMIT_EXCEEDED.NO
+  }
+  const rateLimitResponse = await rateLimitAPI.limit({ key: cid.toString() })
+  if (rateLimitResponse.success) {
+    return RATE_LIMIT_EXCEEDED.NO
+  } else {
+    console.log(`rate limit exceeded: ${cid}`)
+    return RATE_LIMIT_EXCEEDED.YES
+  }
+}
+
+/**
+ * @param {import("../bindings.js").Environment} env
+ * @param {string} authToken
+ * @param {import('@web3-storage/gateway-lib').Context} ctx
+ * @returns {Promise<import('../bindings.js').TokenMetadata | null>}
+ */
+async function getTokenMetadata (env, authToken, ctx) {
+  const cachedValue = await env.AUTH_TOKEN_METADATA.get(authToken)
+  // TODO: we should implement an SWR pattern here - record an expiry in the metadata and if the expiry has passed, re-validate the cache after
+  // returning the value
+  if (cachedValue) {
+    return decode(cachedValue)
```

Comment:
You can `get` and `put` and JSON decode/dnecode so you perhaps don't need these encode/decode functions?:

https://developers.cloudflare.com/kv/api/read-key-value-pairs/#get-method

Reply:
@alanshaw The current API doesn't support that `json` type. I won't change this now because this PR is getting too big, but will certainly update the cloudflare packages in another PR to use the latest version available - which supports `json` values.